### PR TITLE
Make recover work with porous flow

### LINKED ISSUE #12927: PorousFlow recover
## Rationale

I would like PorousFlow to work with "recover".   Requesting help from @cpgr and @hsheldon 

## Description

The following input file, provided by @hsheldon , segfaults:
```
# Testing restart in a hydrothermal simulation using PorousFlow

[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 10
  ny = 1
  nz = 10
  xmin = 0
  xmax = 10000
  ymin = 0
  ymax = 1000
  zmin = -10000
  zmax = 0  
[]

[GlobalParams]
  block = '0'
  PorousFlowDictator = dictator
  gravity = '0 0 -9.81'
[]

[Modules]
  [./FluidProperties]
    [./simple_fluid]
      type = SimpleFluidProperties
      thermal_expansion = 2E-4
      cv = 4180
      cp = 4180
      bulk_modulus = 2E9
      density0 = 1000
      viscosity = 1E-3
    [../]
  [../]
[]

[PorousFlowFullySaturated]
  coupling_type = ThermoHydro
  porepressure = pp
  temperature = temp
  dictator_name = dictator
  fp = simple_fluid
  simulation_type = TRANSIENT
[]

[Variables]
  [./pp]
    [./InitialCondition]
      type = FunctionIC
      function = ini_pp
    [../]
  [../]
  [./temp]
    [./InitialCondition]
      type = FunctionIC
      function = ini_temp
    [../]
  [../]
[]

[Functions]
  [./ini_pp]
    type = ParsedFunction
    vars = 'g rho_f'
    vals = '-10 1000' # grav, fluid dens
    value = 'g*z*rho_f+1E5'
  [../]
  [./ini_temp]
    type = ParsedFunction
    vars = 't_0 t_grad'
    vals = '20 -3e-2' # temp at top, temp gradient
    value = 't_0+t_grad*z'
  [../]
[]

[Kernels]
  # All provided by PorousFlowFullySaturated action
[]

[BCs]
  [./p_top]
    type = FunctionPresetBC
    boundary = 'front'
    variable = pp
    function = ini_pp
  [../]
  [./t_top]
    type = FunctionPresetBC
    boundary = 'front'
    variable = temp
    function = ini_temp
  [../]
  [./t_flux_base]
    type = NeumannBC
    variable = temp
    boundary = 'back'
    value = '70e-3'
  [../]
[]

[AuxVariables]
[]

[AuxKernels]
[]

[Postprocessors]
[]

[Materials]  
  # Thermal conductivity
  [./thermal_conductivity]
    type = PorousFlowThermalConductivityFromPorosity
    lambda_s = '2.9 0 0  0 2.9 0  0 0 2.9'
    lambda_f = '0.65 0 0  0 0.65 0  0 0 0.65'
  [../]  
  
  # Specific heat capacity
  [./rock_heat]
    type = PorousFlowMatrixInternalEnergy
    at_nodes = true
    specific_heat_capacity = 850
    density = 2700
  [../]
  
  # Permeability
  [./permeability]
    type = PorousFlowPermeabilityConst
    permeability = '6E-14 0 0  0 6E-14 0  0 0 6E-14'
  [../]
  
  # Porosity 
  [./porosity_nodal]
    type = PorousFlowPorosityConst
    porosity = 0.1
    at_nodes = true
  [../]

  [./porosity_qp]
    type = PorousFlowPorosityConst
    porosity = 0.1
  [../]

  # Density of saturated rock
  [./density]
    type = PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity
    rho_s = 2700
  [../]
[]

[Preconditioning]
  [./andy]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  solve_type = Newton
  type = Transient
  line_search = bt
  
  dt = 1E11
  end_time = 3E12
  
  l_tol = 1E-5
  nl_abs_tol = 1E-3
  nl_rel_tol = 1E-8
  l_max_its = 200
  nl_max_its = 30
  
  petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -ksp_type -ksp_gmres_restart'
  petsc_options_value = ' asm      2              lu            gmres     200'
[]

[Outputs]
  file_base = test_restart
  exodus = true
  csv = true
  execute_on = 'initial timestep_end'
  checkpoint = true
[]
```

Also, the following tests fail, according to ` ./run_tests --recover`
```
examples/tutorial.10 ................................................................ [recover] FAILED (CRASH)
examples/tutorial.11_2D ............................................................. [recover] FAILED (CRASH)
test:flux_limited_TVD_advection.fltvd_1D_adaptivity ................................ [recover] FAILED (ERRMSG)
test:flux_limited_TVD_advection.fltvd_1D ........................................... [recover] FAILED (ERRMSG)
test:flux_limited_TVD_advection.fltvd_2D_angle ..................................... [recover] FAILED (ERRMSG)
test:flux_limited_TVD_advection.fltvd_2D_blocks .................................... [recover] FAILED (ERRMSG)
test:buckley_leverett.bl01 ......................................................... [recover] FAILED (ERRMSG)
test:flux_limited_TVD_advection.fltvd_2D ........................................... [recover] FAILED (ERRMSG)
test:gravity.grav02a ............................................................... [recover] FAILED (ERRMSG)
test:gravity.grav02d ............................................................... [recover] FAILED (ERRMSG)
test:broadbridge_white.rd03 ......................................................... [recover] FAILED (CRASH)
test:dirackernels.bh07_exo ......................................................... [recover] FAILED (ERRMSG)
test:dirackernels.bh07 ............................................................. [recover] FAILED (ERRMSG)
test:dirackernels.theis2 ........................................................... [recover] FAILED (ERRMSG)
test:dirackernels.theis1 ........................................................... [recover] FAILED (ERRMSG)
test:sinks.s10_bottom ............................................................... [recover] FAILED (CRASH)
test:sinks.s10_back ................................................................. [recover] FAILED (CRASH)
test:sinks.s10_front ................................................................ [recover] FAILED (CRASH)
test:poro_elasticity.terzaghi ....................................................... [recover] FAILED (CRASH)
test:sinks.PorousFlowPiecewiseLinearSink_BC_eg1 ..................................... [recover] FAILED (CRASH)
test:sinks.s10_right ................................................................ [recover] FAILED (CRASH)
test:sinks.injection_production_eg ................................................. [recover] FAILED (ERRMSG)
test:sinks.s10_left ................................................................. [recover] FAILED (CRASH)
test:sinks.s10_top .................................................................. [recover] FAILED (CRASH)
test:poro_elasticity.terzaghi_constM ................................................ [recover] FAILED (CRASH)
test:actions.basicthm_h ............................................................. [recover] FAILED (CRASH)
test:actions.basicthm_hm ............................................................ [recover] FAILED (CRASH)
test:pressure_pulse.pressure_pulse_1d_2phasePS ..................................... [recover] FAILED (ERRMSG)
test:pressure_pulse.pressure_pulse_1d_2phasePSVG ................................... [recover] FAILED (ERRMSG)
test:pressure_pulse.pressure_pulse_1d_2phasePS_KT .................................. [recover] FAILED (ERRMSG)
test:pressure_pulse.pressure_pulse_1d_2phasePSVG2 .................................. [recover] FAILED (ERRMSG)
test:rogers_stallybrass_clements.rsc02 ............................................. [recover] FAILED (ERRMSG)
test:numerical_diffusion.framework ................................................. [recover] FAILED (ERRMSG)
test:numerical_diffusion.fltvd_no_antidiffusion .................................... [recover] FAILED (ERRMSG)
test:numerical_diffusion.fully_saturated_action .................................... [recover] FAILED (ERRMSG)
test:numerical_diffusion.no_action ................................................. [recover] FAILED (ERRMSG)
test:dispersion.diff01 ............................................................. [recover] FAILED (ERRMSG)
test:dispersion.diff01_action ...................................................... [recover] FAILED (ERRMSG)
test:dispersion.disp01 ............................................................. [recover] FAILED (ERRMSG)
test:flux_limited_TVD_pflow.pffltvd_1D ............................................. [recover] FAILED (ERRMSG)
test:flux_limited_TVD_pflow.pffltvd_1D_adaptivity .................................. [recover] FAILED (ERRMSG)
test:flux_limited_TVD_pflow.pffltvd_2D ............................................. [recover] FAILED (ERRMSG)
test:flux_limited_TVD_pflow.pffltvd_2D_angle ....................................... [recover] FAILED (ERRMSG)
test:actions.basicthm_thm ........................................................... [recover] FAILED (CRASH)
test:actions.basicthm_th ................................................... [FINISHED,recover] FAILED (CRASH)
test:numerical_diffusion.fltvd ..................................................... [recover] FAILED (ERRMSG)
```

## Impact

Necessary for users who are using "recover"

### PR DESCRIPTION
**Pull request by cpgr**
Recovery in Porous Flow will only work when we can stop resizing the stateful material data down to the number of quad points when it has been resized to the number of nodes in cases where there are more nodes than quadpoints. This occurs on boundaries where the number of nodes  in the element is greater than the number of quadpoints on the side.

Without this, porous flow seg faults on recovery as the material property storage is sized incorrectly. I don't love this design, but can't think of an easier or better way to do this. It doesn't affect any of the code for other apps, and makes porous flow able to use the checkpoint system for recovery, which is really useful.

Closes #12927

### CODE REVIEW

File: modules/porous_flow/src/materials/PorousFlowMaterial.C [hunk 1]
```diff
@@ -94,17 +86,12 @@ PorousFlowMaterial::sizeNodalProperty(unsigned int prop_id)
    * elements at the end of the std::vector will always be zero, but they
    * are needed because MOOSE does copy operations (etc) that assumes that
    * the std::vector is sized to number of quadpoints.
+   *
+   * On boundary materials, the number of nodes may be larger than the number of
+   * qps on the face of the element, in which case the remaining entries in the
+   * materail properties storage will be zero.
```
Comment (permcody) [inline modules/porous_flow/src/materials/PorousFlowMaterial.C]: s/materail/material/

File: framework/include/materials/MaterialData.h [hunk 2]
```diff
@@ -169,6 +181,10 @@ class MaterialData
   /// Status of storage swapping (calling swap sets this to true; swapBack sets it to false)
   bool _swapped;
 
+  /// Use non-destructive resize of material data (calling resize() will not reduce size).
+  /// Default is false (normal resize behaviour)
+  mutable bool _resize_only_if_smaller;
```
Comment (permcody) [inline framework/include/materials/MaterialData.h]: Why a mutable here? I don't believe MaterialData is constant in the context you are using it nor is the place where you are calling your new onlyResizeIfSmaller() method. Am I missing something?

**Review by permcody — COMMENTED** [review]
This looks OK, but I had a question about your mutable. I'm not sure why this is here or if it's even necessary.

File: framework/include/materials/MaterialData.h (hunk 2 above)
Comment (cpgr) [reply framework/include/materials/MaterialData.h]: I thought that `_material_data` was a const reference but it clearly isn't, so it probably doesn't need to be mutable. I'll remove it

**Comment by cpgr** [comment]
Any chance of a modules recover test in civet to make sure this doesn't break in future?

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
Yes, but probably not near enough. There are several regressions in many of
the apps right now. I'd be willing to setup a target that we could enable
each module one at a time though. What you could do right now is set up a
two part test. You can explicitly pass the "--half-transient" and
"Outputs/checkpoint=true" flags in the first test, then "--recover" in the
second test with a prereq between them. We've done this before in other
apps that aren't ready for full recover testing.

**Comment by WilkAndy** [comment]
Really cool detective work, @cpgr.  Thanks for saving me hours or days by figuring all this out.

 - the changes to the tests, while good changes, weren't anything to do with this PR, were they?  It'd be strange if user's input files could see this code change.

 - can you think about any cases where strain_at_nearest_quadpoint might muck up?

 - can you think of any wierdo element shapes/interpolation schemes/etc where this might fail?  I remember once being perplexed by some sort of tet that had a mismatch between qps and nodes, but can't remember what that was now.

**Comment by cpgr** [comment]
I had to remove the `execute_on = final` etc as it stops the checkpoint files from being written when using `run_tests --recover`. Moving it to a sub-block does work though. I did it for the non-heavy tests in #12929.  Same with adding `initial` to the exodus output. Without it it fails on recover as it tries to open the exodus file again (but it doesn't exist). 

I think it should work even with the strain_at_nearest_quadpoint or weird elements. At least it is working for the usual suspects which is good!

**Comment by cpgr** [comment]
I'll add some two part tests like you suggest @permcody.

**Comment by permcody** [comment]
I've taken several of them out of the framework but you can grep through the test specs for things like "--half-transient" or "--recover" to find the few that remain for other hints on how to get them right. Again, I'm open to starting a recover target for the modules. I guess if you are motivated to get the porous flow tests running with that flag, I'll enable the testing when you are done.

**Pull request merged**